File: NSFRC_UITableView/NSFRC_UITableView/BatchFetchedResultsControllerDelegate.m
//
//  BatchFetchedResultsControllerDelegate.m
//  NSFRC_UITableView
//
//  Created by Heath Borders on 10/9/15.
//  Copyright © 2015 Heath Borders. All rights reserved.
//

#import "BatchFetchedResultsControllerDelegate.h"

@interface BatchFetchedResultsControllerDelegate()

@property (nonatomic) NSMutableArray *deletedRowIndexPaths;
@property (nonatomic) NSMutableArray *insertedRowIndexPaths;
@property (nonatomic) NSMutableArray *updatedRowIndexPaths;

// When building against iOS9, if an entity moves and needs an update,
// we get an update change and then a move change. However, UITableView
// crashes when we do this:
// Without BatchFetchedResultsController (simply mapping change types to their UITableView equivalents)
// *** Assertion failure in -[_UITableViewUpdateSupport _setupAnimationsForExistingVisibleCells], /SourceCache/UIKit_Sim/UIKit-3347.44.2/UITableViewSupport.m:883
// CoreData: error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  Attempt to create two animations for cell with userInfo (null)

// With BatchFetchedResultsController:
// *** Assertion failure in -[UITableView _endCellAnimationsWithContext:], /SourceCache/UIKit_Sim/UIKit-3347.44.2/UITableView.m:1222
// CoreData: error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  attempt to delete and reload the same index path (<NSIndexPath: 0xc000000000000016> {length = 2, path = 0 - 0}) with userInfo (null)
@property (nonatomic) NSIndexPath *maybePreMoveUpdateIndexPath;

@end

@implementation BatchFetchedResultsControllerDelegate

#pragma mark - NSFetchedResultsControllerDelegate

- (void)controller:(NSFetchedResultsController *)controller
   didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath
     forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath {
    void (^abortWithUnexpected)(void) = ^{
        NSLog(@"Unexpected (type, indexPath, newIndexPath): (%@, %@, %@) for object %@ in NSFetchedResultsController: %@",
              @(type),
              indexPath,
              newIndexPath,
              anObject,
              controller);
        abort();
    };
    switch (type) {
        case NSFetchedResultsChangeInsert:
            if (!indexPath && newIndexPath) {
                [self handleMaybeExistingMaybePreMoveUpdate];
                [self.insertedRowIndexPaths addObject:newIndexPath];
            } else {
                abortWithUnexpected();
            }
            break;
        case NSFetchedResultsChangeDelete:
            if (indexPath && !newIndexPath) {
                [self handleMaybeExistingMaybePreMoveUpdate];
                [self.deletedRowIndexPaths addObject:indexPath];
            } else {
                abortWithUnexpected();
            }
            break;
        case NSFetchedResultsChangeMove:
            if (indexPath && newIndexPath) {
                if ([indexPath isEqual:self.maybePreMoveUpdateIndexPath]) {
                    // Since we need to delete and reinsert the row to
                    // properly move it, we don't need to update it.
                    self.maybePreMoveUpdateIndexPath = nil;
                } else {
                    [self handleMaybeExistingMaybePreMoveUpdate];
                }
                // Instead of moving a row around the table, adding and deleting avoids errors that occur when trying to move a row from a deleted section
                [self.deletedRowIndexPaths addObject:indexPath];
                [self.insertedRowIndexPaths addObject:newIndexPath];
            } else {
                abortWithUnexpected();
            }
            break;
        case NSFetchedResultsChangeUpdate:
            if (indexPath && !newIndexPath) {
                self.maybePreMoveUpdateIndexPath = indexPath;
            } else {
                abortWithUnexpected();
            }
            break;
    }
    
}

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    self.deletedRowIndexPaths = [NSMutableArray new];
    self.insertedRowIndexPaths = [NSMutableArray new];
    self.updatedRowIndexPaths = [NSMutableArray new];
    self.maybePreMoveUpdateIndexPath = nil;
}

- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    // there are no more updates, so it wasn't a pre-move update. :)
    [self handleMaybeExistingMaybePreMoveUpdate];
}

#pragma mark - Public API

- (void)clearAfterApplyingToTableView:(UITableView *)tableView {
    [tableView beginUpdates];
    
    [tableView deleteRowsAtIndexPaths:self.deletedRowIndexPaths
                     withRowAnimation:UITableViewRowAnimationLeft];
    [tableView insertRowsAtIndexPaths:self.insertedRowIndexPaths
                     withRowAnimation:UITableViewRowAnimationFade];
    [tableView reloadRowsAtIndexPaths:self.updatedRowIndexPaths
                     withRowAnimation:UITableViewRowAnimationNone];
    
    [tableView endUpdates];
    
    self.deletedRowIndexPaths = nil;
    self.insertedRowIndexPaths = nil;
    self.updatedRowIndexPaths = nil;
    self.maybePreMoveUpdateIndexPath = nil;
}

#pragma mark - Private API

- (void)handleMaybeExistingMaybePreMoveUpdate {
    if (self.maybePreMoveUpdateIndexPath) {
        // there are no more updates, so it wasn't a pre-move update. :)
        [self.updatedRowIndexPaths addObject:self.maybePreMoveUpdateIndexPath];
        self.maybePreMoveUpdateIndexPath = nil;
    }
}

@end

File: TableViewMoveTest/TableViewMoveTest/BatchFetchedResultsControllerDelegate.swift
//
//  BatchFetchedResultsControllerDelegate.swift
//  Pods
//
//  Created by Heath Borders on 6/8/15.
//
//

import Foundation
import CoreData
import UIKit

// adapted from http://www.fruitstandsoftware.com/blog/2013/02/19/uitableview-and-nsfetchedresultscontroller-updates-done-right/
// which is licensed under CC Attribution. Need to figure out what this means for us or if I even need to worry about it:
// https://twitter.com/MrRooni/status/607938115274031104
@objc
public class BatchFetchedResultsControllerDelegate: NSObject {
    
    public override init() {}
    
    // MARK: Public API
    
    public var deletedSectionIndexes: NSIndexSet {
        switch state {
        case .Done(let changeBag):
            return changeBag.deletedSectionIndexes
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public var insertedSectionIndexes: NSIndexSet {
        switch state {
        case .Done(let changeBag):
            return changeBag.insertedSectionIndexes
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public var deletedRowIndexPaths: [NSIndexPath] {
        switch state {
        case .Done(let changeBag):
            return changeBag.deletedRowIndexPaths
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public var insertedRowIndexPaths: [NSIndexPath] {
        switch state {
        case .Done(let changeBag):
            return changeBag.insertedRowIndexPaths
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public var updatedRowIndexPaths: [NSIndexPath] {
        switch state {
        case .Done(let changeBag):
            return changeBag.updatedRowIndexPaths
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public var movedIndexPaths: [NSIndexPath] {
        switch state {
        case .Done(let changeBag):
            return changeBag.movedIndexPaths
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public func clear() {
        switch state {
        case .AwaitingWillChangeContent, .Done:
            state = .AwaitingWillChangeContent
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public func clearAfterApplyingToTableView(tableView: UITableView) {
        switch state {
        case .Done(let changeBag):
            tableView.beginUpdates()
            
            tableView.deleteSections(changeBag.deletedSectionIndexes,
                withRowAnimation: .Left)
            tableView.insertSections(changeBag.insertedSectionIndexes,
                withRowAnimation: .Fade)
            
            tableView.deleteRowsAtIndexPaths(changeBag.deletedRowIndexPaths,
                withRowAnimation: .Left)
            tableView.insertRowsAtIndexPaths(changeBag.insertedRowIndexPaths,
                withRowAnimation: .Fade)
            tableView.reloadRowsAtIndexPaths(changeBag.updatedRowIndexPaths,
                withRowAnimation: .None)
            
            tableView.endUpdates()
            
            clear()
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    public func clearAfterApplyingToCollectionView(collectionView: UICollectionView) {
        switch state {
        case .Done(let changeBag):
            collectionView.performBatchUpdates({
                collectionView.deleteSections(changeBag.deletedSectionIndexes)
                collectionView.insertSections(changeBag.insertedSectionIndexes)
                
                collectionView.deleteItemsAtIndexPaths(changeBag.deletedRowIndexPaths)
                collectionView.insertItemsAtIndexPaths(changeBag.insertedRowIndexPaths)
                collectionView.reloadItemsAtIndexPaths(changeBag.updatedRowIndexPaths)
                },
                completion: .None)
            
            clear()
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    // MARK: Private API
    
    private var state: State = .AwaitingWillChangeContent
}

extension BatchFetchedResultsControllerDelegate: NSFetchedResultsControllerDelegate {
    @objc
    public func controllerWillChangeContent(controller: NSFetchedResultsController) {
        switch state {
        case .AwaitingWillChangeContent, .Done:
            state = .Changing(ChangeBag())
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
    
    @objc
    public func controller(
        controller: NSFetchedResultsController,
        didChangeObject anObject: AnyObject,
        atIndexPath indexPath: NSIndexPath?,
        forChangeType type: NSFetchedResultsChangeType,
        newIndexPath: NSIndexPath?) {
            switch state {
            case .Changing(var changeBag):
                switch (
                    type,
                    indexPath,
                    newIndexPath) {
                case (
                    .Insert,
                    .None,
                    .Some(let newIndexPath)):
                    if changeBag.insertedSectionIndexes.containsIndex(newIndexPath.section) {
                        // If we've already been told that we're adding a section for this inserted row we skip it since it will handled by the section insertion.
                    } else {
                        changeBag.insertedRowIndexPaths.append(newIndexPath)
                    }
                case (
                    .Delete,
                    .Some(let indexPath),
                    .None):
                    if changeBag.deletedSectionIndexes.containsIndex(indexPath.section) {
                        // If we've already been told that we're deleting a section for this deleted row we skip it since it will handled by the section deletion.
                    } else {
                        changeBag.deletedRowIndexPaths.append(indexPath);
                    }
                case (
                    .Move,
                    .Some(let indexPath),
                    .Some(let newIndexPath)):
                    // Instead of moving a row around the table, adding and deleting avoids errors that occur when trying to move a row from a deleted section
                    changeBag.deletedRowIndexPaths.append(indexPath)
                    changeBag.insertedRowIndexPaths.append(newIndexPath)
                case (
                    .Update,
                    .Some(let indexPath),
                    .None):
                    changeBag.updatedRowIndexPaths.append(indexPath)
                default:
                    fatalError("Unexpected (type, indexPath, newIndexPath): (\(type), \(indexPath), \(newIndexPath)) for object: \(anObject) in NSFetchedResultsController: \(controller)")
                }
                
                state = .Changing(changeBag)
            case let illegalState:
                fatalError("Illegal state: \(illegalState)")
            }
    }
    
    @objc
    public func controller(
        controller: NSFetchedResultsController,
        didChangeSection sectionInfo: NSFetchedResultsSectionInfo,
        atIndex sectionIndex: Int,
        forChangeType type: NSFetchedResultsChangeType) {
            switch state {
            case .Changing(let changeBag):
                switch type {
                case .Insert:
                    changeBag.insertedSectionIndexes.addIndex(sectionIndex)
                case .Delete:
                    changeBag.deletedSectionIndexes.addIndex(sectionIndex)
                default:
                    fatalError("Unexpected type: \(type) for sectionInfo: \(sectionInfo) at index: \(sectionIndex) in NSFetchedResultsController: \(controller)")
                }
                
                state = .Changing(changeBag)
            case let illegalState:
                fatalError("Illegal state: \(illegalState)")
            }
    }
    
    @objc
    public func controllerDidChangeContent(controller: NSFetchedResultsController) {
        switch state {
        case .Changing(let changeBag):
            state = .Done(changeBag)
        case let illegalState:
            fatalError("Illegal state: \(illegalState)")
        }
    }
}

private struct ChangeBag {
    let deletedSectionIndexes = NSMutableIndexSet()
    let insertedSectionIndexes = NSMutableIndexSet()
    var deletedRowIndexPaths = [NSIndexPath]()
    var insertedRowIndexPaths = [NSIndexPath]()
    var updatedRowIndexPaths = [NSIndexPath]()
    var movedIndexPaths = [NSIndexPath]()
    
    var totalChanges: Int {
        return
            deletedSectionIndexes.count +
                insertedSectionIndexes.count +
                deletedRowIndexPaths.count +
                insertedRowIndexPaths.count +
                updatedRowIndexPaths.count
    }
}

private enum State {
    case AwaitingWillChangeContent
    case Changing(ChangeBag)
    case Done(ChangeBag)
}
